lib/k8s/conn/auth/exec_worker.ex

```diff
@@ -0,0 +1,215 @@
+      when not is_nil(token) do
+    if valid_token?(state) do
+      {:reply, {:ok, token}, state}
+    else
+      generate_new_token(state)
+    end
+  end
+
+  def handle_call(:get_token, _from, state) do
+    # Generate a new token if the cached token is expired or not present
+    generate_new_token(state)
+  end
+
+  @spec get_token(GenServer.server()) ::
+          {:ok, binary()} | {:error, Jason.DecodeError.t() | Error.t()}
+  def get_token(target) do
+    GenServer.call(target, :get_token)
+  end
+
+  @impl true
+  @spec handle_info(any(), State.t()) :: {:noreply, State.t()}
+  def handle_info(:refresh_token, state) do
+    case response_from_command(state) do
+      {:ok, response} ->
+        new_state = update_state_with_response(state, response) |> schedule_refresh(response)
+        {:noreply, new_state}
+
+      {:error, _error} ->
+        {:noreply, schedule_refresh(state, [])}
+    end
+  end
+
+  @spec generate_new_token(State.t()) ::
+          {:reply, {:ok, binary()}, State.t()} | {:reply, {:error, Error.t()}, State.t()}
+  defp generate_new_token(%State{} = state) do
+    case response_from_command(state) do
+      {:ok, response} ->
+        new_state = update_state_with_response(state, response) |> schedule_refresh(response)
+        {:reply, {:ok, Keyword.get(response, :token)}, new_state}
+
+      {:error, error} ->
+        # Keep the process running
+        {:reply, error, %State{state | token: nil}}
+    end
+  end
+
+  @spec response_from_command(State.t()) ::
+          {:ok, Keyword.t()} | {:error, Jason.DecodeError.t() | Error.t()}
+  defp response_from_command(%{command: command, args: args, env: env} = _config) do
+    # Execute the command by exec'ing the command with the args and env
+    with {cmd_response, 0} <- System.cmd(command, args, env: env),
+         # parse the binary response from the command
+         {:ok, data} <- Jason.decode(cmd_response),
+         # If there's a token in the response, return it
+         {:ok, response} <- parse_cmd_response(data),
+         token when not is_nil(token) <- Keyword.get(response, :token) do
+      {:ok, response}
+    else
+      # If the command fails, return an error
+      {cmd_response, err_code} when is_binary(cmd_response) and is_integer(err_code) ->
+        msg = "#{__MODULE__} failed: #{cmd_response}"
+        {:error, %Error{message: msg}}
+
+      # If there's a parse error or the token is nil, return an error
+      # this just assumes the errors are useful in some way
+      error ->
+        {:error, error}
+    end
+  end
+
+  @spec parse_cmd_response(map) :: {:ok, Keyword.t()} | {:error, Error.t()} | {:error, atom()}
+  defp parse_cmd_response(%{
+         "kind" => "ExecCredential",
+         "status" => %{"token" => token, "expirationTimestamp" => expire}
+       }) do
+    case DateTime.from_iso8601(expire) do
+      {:ok, expiration_timestamp, _} ->
+        {:ok, token: token, expiration_timestamp: expiration_timestamp}
```

Comment:
Does it make sense to return this as a map or tuple? Then we wouldn't have to `Keyword.get` elsewhere.